momepy/preprocessing.py

```diff
@@ -1101,3 +1105,464 @@ def roundabout_simplification(
+    edge_to_col: string
+        edge attribute with the valid destination node id
+
+    Returns
+    ----------
+    Networkx.MultiGraph or Networkx.MultiDiGraph
+
+    """
+    # Collect nodes and their data:
+    nodes, nodes_dict = zip(*graph.nodes(data=True))
+    nodes_df = pd.DataFrame(nodes_dict, index=nodes)
+    nodes_geometries = gpd.points_from_xy(nodes_df[x_col], nodes_df[y_col])
+    graph_crs = graph.graph.get("crs")
+    nodes_gdf = gpd.GeoDataFrame(
+        nodes_df,
+        crs=graph_crs,
+        geometry=nodes_geometries,
+    )
+
+    # In case we did not specify directionality, we infer it from the network:
+    if directed is None:
+        directed = True if isinstance(graph, nx.MultiDiGraph) else False
+
+    # Create a graph without the edges above a certain length and clean it
+    #  from isolated nodes (the unsimplifiable nodes):
+    components_graph = deepcopy(graph)
+    if not directed:
+        components_graph = nx.MultiGraph(components_graph)
+    components_graph.remove_edges_from(
+        [
+            edge
+            for edge in graph.edges(keys=True, data=True)
+            if edge[-1]["length"] > tolerance
+        ]
+    )
+    isolated_nodes_list = list(nx.isolates(components_graph))
+    components_graph.remove_nodes_from(isolated_nodes_list)
+
+    # The connected components of this graph are node clusters we must individually
+    #  simplify. We collect them in a dataframe and retrieve node properties (x, y
+    #  coords mainly) from the original graph.
+    components = nx.connected_components(components_graph)
+    components_dict = dict(enumerate(components, start=max(nodes) + 1))
+    nodes_to_merge_dict = {
+        node: cpt for cpt, nodes in components_dict.items() for node in nodes
+    }
+    new_nodes_df = pd.DataFrame.from_dict(
+        nodes_to_merge_dict, orient="index", columns=["cluster"]
+    )
+    nodes_to_merge_df = pd.concat(
+        [new_nodes_df, nodes_df[[x_col, y_col]]], axis=1, join="inner"
+    )
+
+    # The two node attributes we need for the clusters are the position of the cluster
+    #  centroids. Those are obtained by averaging the x and y columns. We also add
+    # . attribtues referring to the original node ids in every cluster:
+    cluster_centroids_df = nodes_to_merge_df.groupby("cluster").mean()
+    cluster_centroids_df["simplified"] = True
+    cluster_centroids_df["original_node_ids"] = cluster_centroids_df.index.map(
+        components_dict
+    )
+    cluster_geometries = gpd.points_from_xy(
+        cluster_centroids_df[x_col], cluster_centroids_df[y_col]
+    )
+    cluster_gdf = gpd.GeoDataFrame(
+        cluster_centroids_df, crs=graph_crs, geometry=cluster_geometries
+    )
+    cluster_nodes_list = list(cluster_gdf.to_dict("index").items())
+
+    # Rebuild edges if necessary:
+    if rebuild_graph:
+        rebuild_edges_method = rebuild_edges_method.lower()
+        graph.graph["approach"] = "primal"
```

Comment:
Thanks! Wasn't my intention and I didn't catch that. Moved the `simplified_graph` assignment above this block so that I add the approach to that object instead